Make setup_marathon_job dynamically reserve resources

nhandler opened the pull request:
Due to the way we implemented maintenance mode's "draining", it is possible for tasks to get rescheduled onto the at-risk/draining host. To help avoid this and speed up the draining process, we decided to use dynamic reservations. Now, when a host is marked as 'draining', we will create a dynamic reservation for all free resources on the box (preventing anything new from being scheduled there). We will continue to dynamically reserve all free resources each time we kill a task running on the box. Eventually, all resources on the host will be dynamically reserved. The reservations will be removed once the box is undrained.

This change also updates `paasta_metastatus` so that dynamically reserved resources count as `used` resources. This should help us catch issues where we dynamically reserve too many resources in a cluster and no longer have sufficient capacity for new things to run.

I have also filed https://issues.apache.org/jira/browse/MESOS-6175 to try and get some dynamic reservation metrics into the `/metrics/snapshot` endpoint to allow simplifying/optimizing the metastatus code.

Connects to #376

mattmb commented on paasta_tools/mesos_maintenance.py:
```diff
@@ -334,6 +379,118 @@ def load_credentials(mesos_secrets='/nail/etc/mesos-slave-secret'):
+    log.info("Dynamically reserving %f %s on %s" % (amount, resource, slave_id))
+    payload = {
+        'slaveId': slave_id,
+        'resources': str(build_reservation_payload(slave_id, resource, amount)).replace("'", '"').replace('+', '%20')
```
Do we have to do the string replacing? I guess mesos is expecting JSON in a application/x-www-form-urlencoded POST which admittedly seems pretty nasty of it.

What about `json.dumps` and `urllib.quote`. Also shouldn't `+` be `%2B`?

Rob-Johnson replied:
you don't need to do this, requests will urlencode the body for you. just make the value of the ``resources`` key the actual array that ``build_reservation_payload`` returns - requests will take take of the requests

nhandler replied:
@Rob-Johnson That is how I initially had the code. Apparently, while Mesos wants the data submitted as `application/x-www-form-urlencoded`, it parses it like JSON. As a result, it complains if you have single quotes instead of double quotes. `requests` also encodes whitespace as `+`, while Mesos is expecting `%20`.

@mattmb I can try messing around with `urllib.quote`. I hate these `replace()` calls just as much as both of you.

mattmb commented on paasta_tools/mesos_maintenance.py:
```diff
@@ -334,6 +379,118 @@ def load_credentials(mesos_secrets='/nail/etc/mesos-slave-secret'):
+    hosts = []
+    for hostname in hostnames:
+        if '|' in hostname:
+            (host, _) = hostname.split('|')
```
We're repeating this a few times now. I'd suggest splitting it into its own function.

Rob-Johnson replied:
+1, please move this into a function

mattmb commented on paasta_tools/paasta_metastatus.py:
```diff
@@ -204,7 +218,7 @@ def assert_disk_health(metrics, threshold=10):
 
 
-def assert_tasks_running(metrics):
+def assert_tasks_running((metrics, mesos_state)):
```
Accidental search/replace?

Rob-Johnson replied:
no, these are because the ``run_healthchecks_with_param`` expects a single parameter, and these functions now need both the ``mesos_state`` and ``metrics`` params, so @nhandler is wrapping them in a tuple to keep obeying the 'single param' requirement.

This is actually just hinting that the abstraction here is no longer really useful. Long term, we'll get rid of it.

@nhandler my suggestion here would be to drop the usage of ``run_healthchecks_with_param`` from ``get_mesos_metrics_health`` completely. Instead, replace it's implementation with:

```
def get_mesos_metrics_health(mesos_metrics, mesos_state):
    """Perform healthchecks against mesos metrics.
    :param mesos_metrics: a dict exposing the mesos metrics described in
    https://mesos.apache.org/documentation/latest/monitoring/
    :returns: a list of HealthCheckResult tuples
    """
   return [func(mesos_metrics, mesos_state) for func in [
        assert_cpu_health,
        assert_memory_health,
        assert_disk_health,
        assert_tasks_running,
        assert_slave_health,
    ]
```

mattmb replied:
Oh, completely missed the extra bracket!

solarkennedy replied:
I second robj's suggestion

mattmb commented on paasta_tools/paasta_metastatus.py:
```diff
@@ -246,7 +260,7 @@ def assert_no_duplicate_frameworks(state):
 
 
-def assert_slave_health(metrics):
+def assert_slave_health((metrics, mesos_state)):
```
As above?

mattmb commented on paasta_tools/setup_marathon_job.py:
```diff
@@ -242,6 +243,11 @@ def log_bounce_action(line, level='debug'):
+    for task in bounce_lib.flatten_tasks(old_app_at_risk_tasks):
+        if task in tasks_to_kill:
+            hostname = task.host
+            reserve_all_resources([hostname])
```
I'm glad this is only a small addition here 👍  I guess it could be pulled into `kill_given_tasks` which might mean the reservation happens sooner after a task is killed? However, I think we should try it like this first because it reads more cleanly.

solarkennedy replied:
What will un_reserve these resources?

nhandler replied:
The `undrain()` function will call `unreserve_all_resources()`. We will need to make sure things like the autoscaler and other tools remember to call `undrain()`.

solarkennedy replied:
I just wish it wasn't the bounce code's responsibility to do this.

How does this reservation even work if the bounce code hasn't killed these tasks yet?

Also I have a hunch that this needs some try/excepts. The bounce has to go on in the face of slow api endpoints and broken reservation stuff, in the context of reserving away resource for at-risk tasks. Heck, if it was easy I would say make this asynchronous and best-effort so the bounce can move on at normal speed!

nhandler replied:
If the bounce code hasn't killed these tasks yet, it will simply reserve whatever resources are available on the host at the time (which is still better than what it currently does). The next time a task is killed from the host, it will try to reserve all the free resources again. There are other places we can add calls to `reserve_all_resources` (i.e. cron job running on the leader that calls `reserve_all_resources` on any host marked as `draining`), but I'd like to hold off on that for now to avoid hammering the REST endpoints too much and to see how this works in practice.

mattmb commented on paasta_tools/mesos_maintenance.py:
```diff
@@ -334,6 +379,118 @@ def load_credentials(mesos_secrets='/nail/etc/mesos-slave-secret'):
+            cpus = slave['resources']['cpus'] - slave['used_resources']['cpus']
+            for role in slave['reserved_resources']:
+                cpus -= slave['reserved_resources'][role]['cpus']
+            reserve(slave_id=slave_id, resource='cpus', amount=cpus)
```
I guess this could be in a loop if you feel like cutting a few LOC.

Rob-Johnson commented on paasta_tools/mesos_maintenance.py:
```diff
@@ -266,6 +288,29 @@ def get_machine_ids(hostnames):
+            },
+            "role": "maintenance",
+            "reservation": {
+                "principal": str(get_principal()),
```
why're you having to cast to a string here? you explicitly say in ``get_principal`` that it returns a string? might not need to worry about this?

nhandler replied:
This was me fighting against `MagicMocks`. However, I made a number of changes to the itest since then, so let me double check if it still passes without this cast.

nhandler replied:
Yeah, the tests get unhappy without this cast due to the MagicMocks. I've tried fighting them a bit without much luck. I would be more than happy to remove this cast if someone knows of a way to keep the tests passing without it.

Rob-Johnson commented on paasta_tools/mesos_maintenance.py:
```diff
@@ -334,6 +379,118 @@ def load_credentials(mesos_secrets='/nail/etc/mesos-slave-secret'):
+            disk = slave['resources']['disk'] - slave['used_resources']['disk']
+            for role in slave['reserved_resources']:
+                disk -= slave['reserved_resources'][role]['disk']
+            reserve(slave_id=slave_id, resource='disk', amount=disk)
```
can't we bundle each of these up into one request? the endpoint takes an array of resources to reserve, so we can can do all this in one request

Rob-Johnson replied:
further, don't you need to catch the exceptions that might be raised from the HTTP request? we do a ``raise_for_response``, and the docs mention a number of HTTP response codes that might be returned

nhandler replied:
It is because we do a `raise_for_status` in `base_api()` that this becomes a bit more annoying. Since any exception will be caught in this function, which is used by all of the API calls in `mesos_maintenance.py`, we can't easily put error handling specific to dynamic reservations in here.

I think the best approach would be to remove `resp.raise_for_status()` from `base_api()` and move it up two layers into all functions calling `master_api()` (`down`, `up`, `reserve_api`, `unreserve_api`, `maintenance_api`). This will make it possible to adjust how we deal with exceptions for each of the API calls.

Rob-Johnson replied:
@nhandler I don't quite get what you mean. we want to catch exceptions thrown by the ``reserve`` call here, so I was just expecting something like

```
try:
  reserve(slave_id=slave_id, resource='disk', amount=disk)
except HTTPError as e:
   log.debug('couldn't reserve resources; bailing')
   raise
```

(or whatever action you think is the *right* thing to do). [we rereaise the exception after it is caught in](https://github.com/Yelp/paasta/blob/master/paasta_tools/mesos_maintenance.py#L59) ``base_api`` and again in ``reserve`` which I think is fine; you need to catch the HTTPError thrown and deal with it appropriately at each call site.

> This will make it possible to adjust how we deal with exceptions for each of the API calls.

This isn't really useful - the answer is mostly going to be log.debug and raise - the handling of an exception will be unique at each call site, not at the api call

nhandler replied:
Eventually, I think we will want to add some retry logic in the event of a `409 Conflict` (as that will likely be do to race conditions), but for now, I'll simply log and raise.

Rob-Johnson commented on paasta_tools/paasta_metastatus.py:
```diff
@@ -601,7 +619,7 @@ def main():
     )
     metrics = get_mesos_stats()
-    mesos_metrics_status = get_mesos_metrics_health(mesos_metrics=metrics)
+    mesos_metrics_status = get_mesos_metrics_health(mesos_metrics=metrics, mesos_state=mesos_state)
```
We don't have to hit the master state endpoint - we could hit the state endpoint for each slave. It'll mean more work here, as we'll have to send the request to *every* slave in the cluster, and accumulate the results, but it does mean that we can avoid hitting the expensive state endpoint on the master. What do you think?

nhandler replied:
Personally, I think https://issues.apache.org/jira/browse/MESOS-6175 is going to be our best long-term solution. I'm also not sure how well hitting every slave will scale unless we do it in parallel. I also have a feeling we will get hit by timeouts and connectivity issues a bit more frequently, which might slow down this tool. My vote would be to save move this discussion to a separate issue/PR.

Rob-Johnson replied:
sure, but we can't really rely on MESOS-6175. I absolutely agree that we should do this in parallel - in fact, I deleted an unused function from mesos-cli that did this. We can always bring that back to life. Let's see how this goes - I have a feeling we're going to suffer from these calls to the master state.json - but we can ship this version for now

Rob-Johnson commented on paasta_tools/mesos_maintenance.py:
```diff
@@ -334,6 +379,118 @@ def load_credentials(mesos_secrets='/nail/etc/mesos-slave-secret'):
+                unreserve(slave_id=slave_id, resource='disk', amount=disk)
+
+                mem = slave['reserved_resources'][role]['mem']
+                unreserve(slave_id=slave_id, resource='mem', amount=mem)
```
same comments here r.e. bundling into 1 request + error handling

nhandler commented:
I think I've addressed all of the feedback that I have received. However, there were some ugly rebases, so please double check this PR carefully to make sure I didn't do anything stupid by mistake.

solarkennedy commented on paasta_tools/mesos_maintenance.py:
```diff
@@ -331,7 +396,110 @@ def load_credentials(mesos_secrets='/nail/etc/mesos-slave-secret'):
+    log.info("Dynamically reserving resoures on %s: %s" % (slave_id, resources))
+    payload = {
+        'slaveId': slave_id,
+        'resources': str(build_reservation_payload(resources)).replace("'", '"').replace('+', '%20')
```
https://docs.python.org/2/library/urllib.html#urllib.quote ?

nhandler replied:
Nope. The single quotes mess up the parsing and cause an `HTTPError: 400 Client Error: Bad Request`. In my opinion, this is a bug in Mesos itself.

solarkennedy commented on paasta_tools/mesos_maintenance.py:
```diff
@@ -331,7 +396,110 @@ def load_credentials(mesos_secrets='/nail/etc/mesos-slave-secret'):
+    log.info("Dynamically unreserving resoures on %s: %s" % (slave_id, resources))
+    payload = {
+        'slaveId': slave_id,
+        'resources': str(build_reservation_payload(resources)).replace("'", '"').replace('+', '%20')
```
https://docs.python.org/2/library/urllib.html#urllib.quote ?

Rob-Johnson commented on paasta_tools/mesos_maintenance.py:
```diff
@@ -331,7 +399,110 @@ def load_credentials(mesos_secrets='/nail/etc/mesos-slave-secret'):
+
+def reserve_all_resources(hostnames):
+    mesos_state = get_mesos_master().state_summary()
+    hosts = []
```
hosts = [component.host for component in hostnames_to_components(hostnames)]

Rob-Johnson commented on paasta_tools/mesos_maintenance.py:
```diff
@@ -331,7 +399,110 @@ def load_credentials(mesos_secrets='/nail/etc/mesos-slave-secret'):
+    hosts = []
+    components = hostnames_to_components(hostnames)
+    for component in components:
+        hosts.append(component.host)
```
same suggestion as I made in ``reserve_all_resources`` to this logic. The fact that you do this twice hints you should move this logic into its own function with a test

Rob-Johnson replied:
the logic here being duplicated from ``reserve_all_resources`` hints that you should move it into its own function with a unit test

Rob-Johnson commented on paasta_tools/mesos_maintenance.py:
```diff
@@ -331,7 +399,110 @@ def load_credentials(mesos_secrets='/nail/etc/mesos-slave-secret'):
+    components = hostnames_to_components(hostnames)
+    for component in components:
+        hosts.append(component.host)
+    for slave in mesos_state['slaves']:
```
how about removing the inner ``if`` here by filtering those that you're interested in.

```
known_slaves = [slave for slave in mesos_state['slaves'] if slave['hostname'] in hostnames]
for slave in known_slaves:
    log.info("Reserving all resources on %s" % hostname))
    ...
```

Rob-Johnson commented on paasta_tools/mesos_maintenance.py:
```diff
@@ -342,12 +513,13 @@ def drain(hostnames, start, duration):
-    except HTTPError as e:
-        e.msg = "Error performing maintenance drain. Got error: %s" % e.msg
+    except HTTPError:
+        log.debug("Error performing maintenance drain.")
```
why've you made this change? now we lose the extra information in the error message unless we have logging in debug mode?

nhandler replied:
Most (if not all) of these blocks were resulting in their own exceptions being thrown due to `e.msg` not existing. I could probably throw a new `HTTPError` with a more descriptive message rather than simply calling `raise`.

Rob-Johnson replied:
ha, right. yes please!

Rob-Johnson commented on tests/test_mesos_maintenance.py:
```diff
@@ -41,13 +43,18 @@
 from paasta_tools.mesos_maintenance import up
 
 
+Resource = namedtuple('Resource', ['name', 'amount'])
```
why're you redeclaring this?

nhandler replied:
Because I forgot that I could simply import it :)
